[Package]: uv

### Why is it worth to add this package?

Can't install the package normally.

```bash
curl -LsSf https://astral.sh/uv/install.sh | sh
sh: 1034: ldd: not found
System glibc version (`') is too old; using musl
ERROR: there isn't a package for aarch64-linux-android
```

### Home page URL

https://github.com/astral-sh/uv

### Source code URL

https://github.com/astral-sh/uv

### Packaging policy acknowledgement

- [X] The project is actively developed.
- [X] The project has [existing packages](https://repology.org/projects) and is "well known".
- [X] Licensed under an [open source license](https://spdx.org/licenses/).
- [X] Not available through a language package manager: pip, npm, cpan, cargo, etc.

- [X] Not taking up too much disk space (< 100MiB per architecture, exceptions can be made)
- [ ] Not duplicating the functionality of existing packages.
- [X] Not serving hacking, malware, phishing, spamming, spying, ddos functionality.
- [X] I certify that I have read [Termux Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy) and understand that my request will be denied if it is found lacking.

### Additional information

_No response_

I tend to be rather skeptical of any first party benchmarks.
But regardless.

I think this package is *probably* better suited for inclusion in the [TUR](https://github.com/termux-user-repository/tur), since it's essentially `pip` but faster.

---

I can't quite get it building, but it doesn't seem like it would be anything too complicated, just exceeds my knowledge of Rust.

<details><summary>Here's the error I'm getting when it's trying to build the <code>sys-info</code> crate.</summary>
<p>

```styl
error: failed to run custom build command for `sys-info v0.9.1`

Caused by:
  process didn't exit successfully: `/home/builder/.termux-build/python-uv/src/target/release/build/sys-info-42568d11475dc351/build-script-build` (exit status: 1)
  --- stdout
  OUT_DIR = Some(/home/builder/.termux-build/python-uv/src/target/aarch64-linux-android/release/build/sys-info-9532cbfe449f5e48/out)
  TARGET = Some(aarch64-linux-android)
  OPT_LEVEL = Some(3)
  HOST = Some(x86_64-unknown-linux-gnu)
  cargo:rerun-if-env-changed=CC_aarch64-linux-android
  CC_aarch64-linux-android = None
  cargo:rerun-if-env-changed=CC_aarch64_linux_android
  CC_aarch64_linux_android = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = Some(aarch64-linux-android-clang)
  RUSTC_WRAPPER = None
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some(false)
  cargo:rerun-if-env-changed=CFLAGS_aarch64-linux-android
  CFLAGS_aarch64-linux-android = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64_linux_android
  CFLAGS_aarch64_linux_android = Some( -I/data/data/com.termux/files/usr/include)
  cargo:warning=c/linux.c:96:11: error: call to undeclared library function 'index' with type 'char *(const char *, int)'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  cargo:warning=   96 |                 mount = index(procline, ' ');
  cargo:warning=      |                         ^
  cargo:warning=c/linux.c:96:11: note: include the header <strings.h> or explicitly provide a declaration for 'index'
  cargo:warning=1 error generated.

  --- stderr


  error occurred: Command "aarch64-linux-android-clang" "-O3" "-DANDROID" "-ffunction-sections" "-fdata-sections" "-fPIC" "-I/data/data/com.termux/files/usr/include" "-o" "/home/builder/.termux-build/python-uv/src/target/aarch64-linux-android/release/build/sys-info-9532cbfe449f5e48/out/db3b6bfb95261072-linux.o" "-c" "c/linux.c" with args aarch64-linux-android-clang did not execute successfully (status code exit status: 1).


warning: build failed, waiting for other jobs to finish...
```

</p>
</details> 

And here's the build script.
```bash
# file: tur/python-uv/build.sh
TERMUX_PKG_HOMEPAGE=https://github.com/astral-sh/uv
TERMUX_PKG_DESCRIPTION="An extremely fast Python package installer and resolver, written in Rust."
TERMUX_PKG_LICENSE="Apache-2.0, MIT"
TERMUX_PKG_LICENSE_FILE="LICENSE-APACHE, LICENSE-MIT"
TERMUX_PKG_MAINTAINER="@termux-user-repository"
TERMUX_PKG_VERSION=0.2.35
TERMUX_PKG_SRCURL=https://github.com/astral-sh/uv/archive/refs/tags/${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_SHA256=75233b4fe20715ac2da9e14e35379ed67f025e267fbe63abcb5d607033572c41
TERMUX_PKG_BUILD_DEPENDS='libandroid-support, libbz2, zlib'
TERMUX_PKG_DEPENDS="python"
TERMUX_PKG_PYTHON_COMMON_DEPS="'setuptools>=65.0.0', wheel"
TERMUX_PKG_BUILD_IN_SRC=true
TERMUX_PKG_AUTO_UPDATE=true
TERMUX_PKG_UPDATE_TAG_TYPE=latest-release-tag

termux_step_pre_configure() {
	termux_setup_cmake
	termux_setup_rust

	# borrowed from https://github.com/termux/termux-packages/blob/55254de403f139b508cd9fb392c2082b72928a8b/packages/fselect/build.sh#L15-L21
	# Dummy CMake toolchain file to workaround build error:
	# error: failed to run custom build command for `libz-ng-sys v1.1.15`
	# ...
	# CMake Error at /home/builder/.termux-build/_cache/cmake-3.28.3/share/cmake-3.28/Modules/Platform/Android-Determine.cmake:217 (message):
	# Android: Neither the NDK or a standalone toolchain was found.
	export TARGET_CMAKE_TOOLCHAIN_FILE="${TERMUX_PKG_BUILDDIR}/android.toolchain.cmake"
	touch "${TERMUX_PKG_BUILDDIR}/android.toolchain.cmake"
}

termux_step_make() {
	cargo build --jobs $TERMUX_PKG_MAKE_PROCESSES --target $CARGO_TARGET_NAME --release
}

termux_step_make_install() {
	install -Dm755 -t $TERMUX_PREFIX/bin target/${CARGO_TARGET_NAME}/release/uv
}
```

--- a/packages/uv/0001-sys-info-replace-index-with-strchr.diff
+++ b/packages/uv/0001-sys-info-replace-index-with-strchr.diff
@@ -1,0 +1,36 @@
+https://github.com/FillZpp/sys-info-rs/pull/118
+From e68597ffd323107bc766e6b04e565fd907d9820a Mon Sep 17 00:00:00 2001
+From: bb010g <me@bb010g.com>
+Date: Mon, 13 May 2024 06:57:39 -0700
+Subject: [PATCH] fix: `index()` -> `strchr()`
+
+`index()` is deprecated.
+---
+ c/linux.c | 8 ++++----
+ 1 file changed, 4 insertions(+), 4 deletions(-)
+
+diff --git a/c/linux.c b/c/linux.c
+index 4cd1589..ede98d5 100644
+--- a/c/linux.c
++++ b/c/linux.c
+@@ -93,16 +93,16 @@ DiskInfo get_disk_info(void) {
+ 	}
+ 	while ( fgets(procline, sizeof(procline), mounts) ) {
+ 		device = procline;
+-		mount = index(procline, ' ');
++		mount = strchr(procline, ' ');
+ 		if (mount == NULL) continue;
+ 		*mount++ = '\0';
+-		type = index(mount, ' ');
++		type = strchr(mount, ' ');
+ 		if (type == NULL) continue;
+ 		*type++ = '\0';
+-		mode = index(type, ' ');
++		mode = strchr(type, ' ');
+ 		if (mode == NULL) continue;
+ 		*mode++ = '\0';
+-		other = index(mode, ' ');
++		other = strchr(mode, ' ');
+ 		if (other != NULL) *other = '\0';
+ 		if (!strncmp(mode, "ro", 2)) continue;
+ 		if (remote_mount(device, type)) continue;

--- a/packages/uv/build.sh
+++ b/packages/uv/build.sh
@@ -1,0 +1,71 @@
+TERMUX_PKG_HOMEPAGE=https://docs.astral.sh/uv/
+TERMUX_PKG_DESCRIPTION="An extremely fast Python package installer and resolver, written in Rust."
+TERMUX_PKG_LICENSE="Apache-2.0, MIT"
+TERMUX_PKG_LICENSE_FILE="LICENSE-APACHE, LICENSE-MIT"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION="0.3.2"
+TERMUX_PKG_SRCURL=https://github.com/astral-sh/uv/archive/refs/tags/${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=f81ecd38dbaad3664084549a7f2152b6042a1984e2cee3b0d5f90751a8dc9ad3
+TERMUX_PKG_BUILD_DEPENDS="zstd"
+TERMUX_PKG_BUILD_IN_SRC=true
+TERMUX_PKG_AUTO_UPDATE=true
+
+termux_step_pre_configure() {
+	termux_setup_cmake
+	termux_setup_rust
+
+	# Dummy CMake toolchain file to workaround build error:
+	# error: failed to run custom build command for `libz-ng-sys v1.1.15`
+	# ...
+	# CMake Error at /home/builder/.termux-build/_cache/cmake-3.28.3/share/cmake-3.28/Modules/Platform/Android-Determine.cmake:217 (message):
+	# Android: Neither the NDK or a standalone toolchain was found.
+	export TARGET_CMAKE_TOOLCHAIN_FILE="${TERMUX_PKG_BUILDDIR}/android.toolchain.cmake"
+	touch "${TERMUX_PKG_BUILDDIR}/android.toolchain.cmake"
+
+	: "${CARGO_HOME:=$HOME/.cargo}"
+	export CARGO_HOME
+
+	rm -rf "${CARGO_HOME}"/registry/src/*/sys-info-*
+	cargo fetch --target "${CARGO_TARGET_NAME}"
+
+	patch -p1 -d "${CARGO_HOME}"/registry/src/*/sys-info-* \
+	-i "${TERMUX_PKG_BUILDER_DIR}"/0001-sys-info-replace-index-with-strchr.diff
+}
+
+termux_step_make() {
+	PKG_CONFIG_ALL_DYNAMIC=1 \
+	ZSTD_SYS_USE_PKG_CONFIG=1 \
+	cargo build --jobs "${TERMUX_PKG_MAKE_PROCESSES}" --target "${CARGO_TARGET_NAME}" --release
+}
+
+termux_step_make_install() {
+	install -Dm700 -t "${TERMUX_PREFIX}"/bin target/"${CARGO_TARGET_NAME}"/release/uv
+	install -Dm700 -t "${TERMUX_PREFIX}"/bin target/"${CARGO_TARGET_NAME}"/release/uvx
+}
+
+termux_step_post_make_install() {
+	# Make a placeholder for shell-completions (to be filled with postinst)
+	mkdir -p "${TERMUX_PREFIX}"/share/bash-completion/completions
+	mkdir -p "${TERMUX_PREFIX}"/share/elvish/lib
+	mkdir -p "${TERMUX_PREFIX}"/share/fish/vendor_completions.d
+	mkdir -p "${TERMUX_PREFIX}"/share/zsh/site-functions
+	touch "${TERMUX_PREFIX}"/share/bash-completion/completions/uv
+	touch "${TERMUX_PREFIX}"/share/elvish/lib/uv.elv
+	touch "${TERMUX_PREFIX}"/share/fish/vendor_completions.d/uv.fish
+	touch "${TERMUX_PREFIX}"/share/zsh/site-functions/_uv
+}
+
+termux_step_post_massage() {
+	rm -rf "${CARGO_HOME}"/registry/src/*/sys-info-*
+}
+
+termux_step_create_debscripts() {
+	cat <<-EOF >./postinst
+		#!${TERMUX_PREFIX}/bin/sh
+
+		uv generate-shell-completion bash > "${TERMUX_PREFIX}/share/bash-completion/completions/uv"
+		uv generate-shell-completion elvish > "$TERMUX_PREFIX/share/elvish/lib/uv.elv"
+		uv generate-shell-completion fish > "${TERMUX_PREFIX}/share/fish/vendor_completions.d/uv.fish"
+		uv generate-shell-completion zsh > "${TERMUX_PREFIX}/share/zsh/site-functions/_uv"
+	EOF
+}

